### ISSUE TITLE
[red-knot] Protocol method return type is always inferred as `None`

### ISSUE THREAD

**Issue by jorenham** [astral-sh/ruff#16895]
### Summary

I'm not sure if I should call this a bug or a missing feature, but either way, red-knot is not behaving according to the typing spec here:

```py
from typing import Protocol

class CanSpam[OutT](Protocol):
    def spam(self, /) -> OutT: ...
```

which reports

```
Function can implicitly return `None`, which is not assignable to return type `OutT` (lint:invalid-return-type) [Ln 4, Col 26]
```

https://playknot.ruff.rs/5d3edee7-5070-4046-a34a-6c1b8aa2340d

I'm sure can see why I think it's not behaving as it should here. If not, then I'd be happy to explain it <sub>but be aware that that will probably involve cringy examples, awkward puns, unnecessary jargon, and a non-negligible chance at me side-tracking and bursting out into a rant about something completely unrelated</sub> 👌🏻.

**Comment by AlexWaygood** [comment]
Yes, we need to special-case (at least) three kinds of methods and exclude them from this check:
- Methods in classes that inherit directly from `typing.Protocol`
- Methods decorated with `@abc.abstractmethod`
- Methods decorated with `@typing.overload`

**Comment by jorenham** [comment]
> Yes, we need to special-case (at least) three kinds of functions and exclude them from this check:
> 
> * Methods in classes that inherit directly from `typing.Protocol`
> * Methods decorated with `@abc.abstractmethod`
> * Functions or methods decorated with `@typing.overload`

I suppose function in a `*.pyi` would also fit this bill
